### PR TITLE
fix: lower default pricing and check LP balance

### PR DESCRIPTION
**Pull request by alvin-reyes**
### Summary

This pull request makes the following changes:

- [ ] Lower the instruction price to add resource offers
- [ ] Add a LP balance check 

### Task/Issue reference

Closes: [add_link_here](https://github.com/Lilypad-Tech/internal/issues/309)

### Test plan

Please describe how reviewers can test the changes in this pull request. If the test plan is challenging to explain in text alone, a short video demonstration may be included.

#### Positive tests
- Run the RP with an address ETH balance of > 0.0006 ETH 
- Run the RP with an address LP balance of > 0 LP
- Run the `GetBalance` and `GetLPBalance` unit tests

#### Negative tests
- Run the RP with an address ETH balance of 0 OR < 0.0006 ETH 
- Run the RP with an address LP balance of 0 or < 1 LP

## More details

- Key thing to note here is that we will need to add a new DealPricing property for LP and ETH as part of the job fee payment. I'll be putting adding that on a separate PR.

- The 0.0006 requirement here is based on the fees we pay per job. It is not an AVG, but just a educated guess. I'll have to create a new way of getting the fees (through indexing). Possibly a microservice to get the past 100 blocks, average the fee and use the value instead. I'll put this on a separate PR.

### CODE REVIEW

File: pkg/solver/controller.go [hunk 1]
```diff
@@ -332,12 +332,23 @@ func (controller *SolverController) addResourceOffer(resourceOffer data.Resource
 		return nil, fmt.Errorf("failed to retrieve ETH balance for resource provider: %v", err)
 	}
 	// Convert InstructionPrice from ETH to Wei
-	requiredBalanceWei := web3.EtherToWei(float64(resourceOffer.DefaultPricing.InstructionPrice))
+	requiredBalanceWei := web3.EtherToWei(0.0006) // 0.0006 based on the required balance for a job
```
Comment (narbs91) [inline pkg/solver/controller.go]: [nit] Make the `0.0006` a variable so it's easier to track and change when we need to

File: pkg/web3/sdk.go [hunk 2]
```diff
@@ -308,3 +312,30 @@ func (sdk *Web3SDK) GetBalance(address string) (*big.Int, error) {
...
+	erc20ABIObj, err := abi.JSON(strings.NewReader(erc20ABI))
+	if err != nil {
+		log.Printf("error parsing ABI: %s", err)
+	}
+	tokenContract := bind.NewBoundContract(lpToken, erc20ABIObj, client, client, client)
+
+	var out []interface{}
+	err = tokenContract.Call(nil, &out, "balanceOf", ethAddress)
+	if err != nil {
+		log.Printf("error calling balanceOf: %s", err)
```
Comment (narbs91) [inline pkg/web3/sdk.go]:
```suggestion
		log.Error().Msgf("error calling balanceOf: %s", err)
```

File: pkg/web3/sdk.go [hunk 3]
```diff
@@ -308,3 +312,30 @@ func (sdk *Web3SDK) GetBalance(address string) (*big.Int, error) {
...
+func (sdk *Web3SDK) GetLPBalance(address string) (*big.Int, error) {
+	// Convert the string address to common.Address
+	client := sdk.Client
+	ethAddress := common.HexToAddress(address)
+	lpToken := common.HexToAddress(sdk.Options.TokenAddress) // LP Token Address
+
+	// Get the balance using the converted address
+	erc20ABIObj, err := abi.JSON(strings.NewReader(erc20ABI))
+	if err != nil {
+		log.Printf("error parsing ABI: %s", err)
```
Comment (narbs91) [inline pkg/web3/sdk.go]:
```suggestion
		log.Error().Msgf("error parsing ABI: %s", err)
```

File: pkg/web3/sdk_test.go [hunk 4]
```diff
@@ -70,6 +73,18 @@ func TestGetBalance(t *testing.T) {
 	t.Logf("Balance: %d\n", balance)
 }
 
+func TestGetLPBalance(t *testing.T) {
```
Comment (narbs91) [inline pkg/web3/sdk_test.go:80]: Maybe let's add another test for the case where there the address doesn't have enough LP

**Review by narbs91 — COMMENTED** [review]
LGTM, just some minor comments

File: pkg/web3/sdk.go (hunk 3 above)
Comment (alvin-reyes) [reply pkg/web3/sdk.go]: resolved on the latest commit.

File: pkg/web3/sdk_test.go (hunk 4 above)
Comment (alvin-reyes) [reply pkg/web3/sdk_test.go:80]: resolved on the latest commit.

File: pkg/web3/sdk.go (hunk 2 above)
Comment (alvin-reyes) [reply pkg/web3/sdk.go]: resolved on the latest commit.

File: pkg/solver/controller.go (hunk 1 above)
Comment (alvin-reyes) [reply pkg/solver/controller.go]: resolved on the latest commit.

File: pkg/web3/sdk.go [hunk 5]
```diff
@@ -308,3 +312,30 @@ func (sdk *Web3SDK) GetBalance(address string) (*big.Int, error) {
...
+		log.Error().Msgf("error parsing ABI: %s", err)
+	}
+	tokenContract := bind.NewBoundContract(lpToken, erc20ABIObj, client, client, client)
+
+	var out []interface{}
+	err = tokenContract.Call(nil, &out, "balanceOf", ethAddress)
+	if err != nil {
+		log.Error().Msgf("error calling balanceOf: %s", err)
+	}
+	lpBalance := *abi.ConvertType(out[0], new(big.Int)).(*big.Int)
```
Comment (narbs91) [inline pkg/web3/sdk.go:337]: I think the 'out[0]' will throw and out of bounds error here if we enter the err block above it.  Should we return if we're in the error block just above this line?

Comment (alvin-reyes) [reply pkg/web3/sdk.go:337]: good catch. we should. resolved on the latest commit.

**Review by narbs91 — APPROVED** [review]

**Pull request merged**